### PR TITLE
Auto-exclude conflicting Spring Boot security auto-configurations

### PR DESCRIPTION
**Pull request by jamesfredley**
## Summary

Adds `SecurityAutoConfigurationExcluder` implementing `AutoConfigurationImportFilter` to automatically exclude 7 Spring Boot security auto-configuration classes that conflict with the Grails Spring Security plugin. This eliminates the manual `spring.autoconfigure.exclude` entries that every Grails 7 user currently must add to `application.yml`.

## Feature Description

The plugin [README](https://github.com/apache/grails-spring-security/blob/7.0.x/README.md#L53-L66) documents that Grails 7 requires 7 manual `spring.autoconfigure.exclude` entries:

```yaml
spring:
  autoconfigure:
    exclude:
      - org.springframework.boot.autoconfigure.security.servlet.SecurityAutoConfiguration
      - org.springframework.boot.autoconfigure.security.servlet.SecurityFilterAutoConfiguration
      - org.springframework.boot.autoconfigure.security.servlet.UserDetailsServiceAutoConfiguration
      - org.springframework.boot.autoconfigure.security.oauth2.client.servlet.OAuth2ClientAutoConfiguration
      - org.springframework.boot.autoconfigure.security.oauth2.client.OAuth2ClientAutoConfiguration
      - org.springframework.boot.autoconfigure.security.oauth2.resource.servlet.OAuth2ResourceServerAutoConfiguration
      - org.springframework.boot.actuate.autoconfigure.security.servlet.ManagementWebSecurityAutoConfiguration
```

These exclusions are **always required** — not conditional or environment-specific. This boilerplate is easy to miss, hard to debug when forgotten, and should be handled automatically by the plugin.

## Implementation

**Approach**: `AutoConfigurationImportFilter` SPI (stable since Spring Boot 1.5.0, used in 3.x)

### Why `AutoConfigurationImportFilter` over alternatives

| Approach | Drawback |
|----------|----------|
| `EnvironmentPostProcessor` | `spring.autoconfigure.exclude` is a `List` — property sources cannot merge lists reliably ([Spring Boot #41669](https://github.com/spring-projects/spring-boot/issues/41669)). User's own exclusions would be overwritten. |
| `@EnableAutoConfiguration(exclude=...)` on plugin class | Plugin uses `@Grails` annotation, can't also use `@EnableAutoConfiguration` |
| Documentation-only (current approach) | Every user must copy 7 YAML lines; easy to miss |
| **`AutoConfigurationImportFilter`** | **✓ Runs before bytecode is loaded (faster)**<br>**✓ Opt-out via `excludeSpringSecurityAutoConfiguration: false`**<br>**✓ No property merging issues**<br>**✓ Used by established libraries (Redis OM Spring, TCC Transaction)** |

### Files Changed

| File | Change |
|------|--------|
| `SecurityAutoConfigurationExcluder.groovy` | **NEW** — `AutoConfigurationImportFilter` implementation that returns `false` for the 7 conflicting auto-configurations |
| `META-INF/spring.factories` | **NEW** — Registers the filter via SPI (filters still use `spring.factories` in Spring Boot 3.x, not `.imports`) |
| `build.gradle` | Added `compileOnly 'org.springframework.boot:spring-boot-autoconfigure'` — always available at runtime in any Grails app |
| `SecurityAutoConfigurationExcluderSpec.groovy` | **NEW** — 18 Spock tests covering all exclusions, preservation of non-security configs, mixed arrays, edge cases, and `spring.factories` registration |

### Test Coverage (18 tests, all pass)

- 7 data-driven tests: each excluded auto-configuration is filtered out
- 5 data-driven tests: non-security auto-configurations pass through
- 1 test: mixed array of included/excluded classes
- 1 test: empty array
- 1 test: null metadata parameter
- 1 test: all 7 known classes present in the exclusion set
- 1 test: exclusion set is unmodifiable
- 1 test: `spring.factories` registration is correct

### Backward Compatibility

- Users who already have manual exclusions in their `application.yml` are unaffected — the filter and manual exclusions are independent mechanisms
- No behavioral change for existing functionality
- `spring-boot-autoconfigure` is added as `compileOnly` only — no new runtime dependency (it's always already present via Spring Boot starter)

## Example Application

https://github.com/jamesfredley/grails-spring-security-autoconfig-exclusion

A minimal Grails 7.0.7 app with spring-security configured (User/Role/UserRole domains, annotation-based security) that intentionally does NOT include the manual exclusions. The `/bugDemo/index` endpoint shows which auto-configuration classes are on the classpath and which beans are registered.

## Environment Information

- **Grails**: 7.0.7
- **Spring Boot**: 3.5.10
- **Groovy**: 4.0.30
- **JDK**: 17

## Version

7.0.x

## Opt-Out

Users can disable the automatic exclusion by setting the following property in `application.yml`:

```yaml
grails:
  plugin:
    springsecurity:
      excludeSpringSecurityAutoConfiguration: false
```

This allows Spring Boot's security auto-configurations to run normally, for users who want to override or use Spring Boot's security setup directly. The filter implements `EnvironmentAware` to read this property during the auto-configuration discovery phase.

### CODE REVIEW

File: plugin-core/plugin/src/test/groovy/grails/plugin/springsecurity/SecurityAutoConfigurationExcluderSpec.groovy [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+        then: 'the resource exists'
+        resource != null
+
+        when:
+        String content = resource.text
+
+        then: 'it registers SecurityAutoConfigurationExcluder as an AutoConfigurationImportFilter'
+        content.contains('org.springframework.boot.autoconfigure.AutoConfigurationImportFilter')
+        content.contains('grails.plugin.springsecurity.SecurityAutoConfigurationExcluder')
```
[the remark this answers is not in the document]
Comment (jamesfredley) [reply plugin-core/plugin/src/test/groovy/grails/plugin/springsecurity/SecurityAutoConfigurationExcluderSpec.groovy]: Good catch — fixed. Switched from `getResource()` to `getResources()` (plural) and now scans all `META-INF/spring.factories` files on the classpath via `Collections.list(resources)`, then asserts that at least one contains our registration. This avoids classpath ordering flakiness.

File: plugin-core/plugin/src/main/groovy/grails/plugin/springsecurity/SecurityAutoConfigurationExcluder.groovy [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+     *   <li>{@code UserDetailsServiceAutoConfiguration} — creates an in-memory
+     *       {@code UserDetailsService} that conflicts with the plugin's
+     *       {@code GormUserDetailsService}</li>
+     *   <li>{@code OAuth2ClientAutoConfiguration} (servlet) — conflicts when the
+     *       plugin-oauth2 module manages OAuth2 configuration</li>
+     *   <li>{@code OAuth2ResourceServerAutoConfiguration} — conflicts with the
+     *       plugin's resource server security setup</li>
+     *   <li>{@code ManagementWebSecurityAutoConfiguration} — Actuator security
+     *       that conflicts when Actuator is on the classpath</li>
+     * </ul>
```
[the remark this answers is not in the document]
Comment (jamesfredley) [reply plugin-core/plugin/src/main/groovy/grails/plugin/springsecurity/SecurityAutoConfigurationExcluder.groovy:95]: Good catch — fixed. The Javadoc now documents all 7 entries, with the two `OAuth2ClientAutoConfiguration` variants listed separately (`...oauth2.client.servlet` and `...oauth2.client`).

**Comment by jamesfredley** [comment]
related to https://github.com/apache/grails-core/pull/15405, which needs exclusions also

**Review by jdaugherty — DISMISSED** [review]
How do you opt out of this?  What if you want to override it?

**Comment by jamesfredley** [comment]
Addressed - the filter now implements `EnvironmentAware` and checks a property to allow opting out:

```yml
grails:
  plugin:
    springsecurity:
      excludeSpringSecurityAutoConfiguration: false
```

Defaults to `true` (exclusions active). Setting it to `false` disables the filter entirely and lets all Spring Boot security auto-configurations through - for users who want to override or use Spring Boot's security setup directly.

Also updated the README to document both the automatic behavior and the opt-out.

File: plugin-core/plugin/src/main/groovy/grails/plugin/springsecurity/SecurityAutoConfigurationExcluder.groovy [hunk 3]
```diff
@@ -38,6 +40,16 @@ import org.springframework.boot.autoconfigure.AutoConfigurationMetadata
  * automatically filtering them out during Spring Boot's auto-configuration
  * discovery phase.</p>
  *
+ * <p>To disable this filter and allow Spring Boot's security auto-configurations
```
Comment (jdaugherty) [inline plugin-core/plugin/src/main/groovy/grails/plugin/springsecurity/SecurityAutoConfigurationExcluder.groovy:44]: I believe we need to put this in the actual spring docs as well as the groovydoc.

File: plugin-core/plugin/src/main/groovy/grails/plugin/springsecurity/SecurityAutoConfigurationExcluder.groovy [hunk 4]
```diff
@@ -46,7 +58,16 @@ import org.springframework.boot.autoconfigure.AutoConfigurationMetadata
...
-class SecurityAutoConfigurationExcluder implements AutoConfigurationImportFilter {
+class SecurityAutoConfigurationExcluder implements AutoConfigurationImportFilter, EnvironmentAware {
+
+    static final String ENABLED_PROPERTY = 'grails.plugin.springsecurity.excludeSpringSecurityAutoConfiguration'
+
+    private boolean enabled = true
+
+    @Override
+    void setEnvironment(Environment environment) {
+        this.enabled = environment.getProperty(ENABLED_PROPERTY, Boolean, true)
```
Comment (jdaugherty) [inline plugin-core/plugin/src/main/groovy/grails/plugin/springsecurity/SecurityAutoConfigurationExcluder.groovy:70]: I don't think this will work without my plugin fix.  Maybe if it's specified in the end app it will, but I don't see any ordering on this excluder.

File: plugin-core/plugin/src/test/groovy/grails/plugin/springsecurity/SecurityAutoConfigurationExcluderSpec.groovy [hunk 5]
```diff
@@ -147,6 +148,59 @@ class SecurityAutoConfigurationExcluderSpec extends Specification {
         thrown(UnsupportedOperationException)
     }
 
+    def "match allows all auto-configurations when disabled via environment property"() {
```
Comment (jdaugherty) [inline plugin-core/plugin/src/test/groovy/grails/plugin/springsecurity/SecurityAutoConfigurationExcluderSpec.groovy:152]: I think you should an actual functional test for this.  I'm pretty sure this coudl be broken, but it may only be in edge cases with plugin's having configuration instead of the end app.

**Review by jdaugherty — DISMISSED** [review]

File: plugin-core/plugin/src/main/groovy/grails/plugin/springsecurity/SecurityAutoConfigurationExcluder.groovy (hunk 4 above)
Comment (jamesfredley) [reply plugin-core/plugin/src/main/groovy/grails/plugin/springsecurity/SecurityAutoConfigurationExcluder.groovy:70]: This does work. Spring Boot's `AutoConfigurationImportSelector` explicitly calls `invokeAwareMethods(filter)` on every `AutoConfigurationImportFilter` **before** `match()` is ever called. The `invokeAwareMethods()` method explicitly handles `EnvironmentAware`:


Test app: https://github.com/jamesfredley/grails-spring-security-autoconfig-exclusion

File: plugin-core/plugin/src/test/groovy/grails/plugin/springsecurity/SecurityAutoConfigurationExcluderSpec.groovy (hunk 5 above)
Comment (jamesfredley) [reply plugin-core/plugin/src/test/groovy/grails/plugin/springsecurity/SecurityAutoConfigurationExcluderSpec.groovy:152]: Added in `9656c948`:

1. **Integration test in the plugin repo** (`SecurityAutoConfigurationExcluderIntegrationSpec`) - runs in the `core-examples-integration-test-app` with the full Grails context. Verifies the excluder is on the classpath, `SecurityAutoConfiguration` and `SecurityFilterAutoConfiguration` beans are NOT registered, no duplicate `SecurityFilterChain` beans, and only the plugin's `UserDetailsService` is active. All 5 assertions pass.

2. **External sample app** with two integration test specs:
   - `SecurityAutoConfigExclusionSpec` (6 tests) - proves auto-configs are excluded by default
   - `SecurityAutoConfigOptOutDisabledSpec` (1 test) - proves setting `excludeSpringSecurityAutoConfiguration: false` lets `SecurityAutoConfiguration` through (confirming the `EnvironmentAware` opt-out works)

   App: https://github.com/jamesfredley/grails-spring-security-autoconfig-exclusion (uses locally-built `7.0.2-SNAPSHOT` from this PR branch via `mavenLocal()`)

File: plugin-core/plugin/src/main/groovy/grails/plugin/springsecurity/SecurityAutoConfigurationExcluder.groovy (hunk 4 above)
Comment (jdaugherty) [reply plugin-core/plugin/src/main/groovy/grails/plugin/springsecurity/SecurityAutoConfigurationExcluder.groovy:70]: I'm more worried about the configuration being picked up if it's defined in a plugin.  It probably works on the end app, so that's probably acceptalbe for now.

File: plugin-core/plugin/src/test/groovy/grails/plugin/springsecurity/SecurityAutoConfigurationExcluderSpec.groovy (hunk 5 above)
Comment (jdaugherty) [reply plugin-core/plugin/src/test/groovy/grails/plugin/springsecurity/SecurityAutoConfigurationExcluderSpec.groovy:152]: we have example apps in the srpign security repo, why not import it here?

**Review by jdaugherty — APPROVED** [review]
I'm glad to see it working, wecan merge this.  I'd prefer we just import a functional test app here though.

**Review by matrei — APPROVED** [review]

**Comment by matrei** [comment]
This is great!
